velox/vector/VectorPool.h

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include "velox/vector/FlatVector.h"
+
+namespace facebook::velox {
+
+// A thread-level cache of preallocated flat vectors of different types.
+class VectorPool {
+ public:
+  VectorPtr
+  get(const TypePtr& type, vector_size_t size, memory::MemoryPool& pool) {
+    auto kind = static_cast<int32_t>(type->kind());
+    if (kind < kNumCachedVectorTypes && size <= kMaxRecycleSize) {
+      return vectors_[kind].pop(type, size, pool);
+    }
+    return BaseVector::create(type, size, &pool);
+  }
+
+  // Moves vector into 'this' if it is flat, recursively singly referenced and
+  // there is space.
+  void release(VectorPtr& vector) {
```

Comment:
How about taking `VectorPtr&&` to make it clear that the pointer is not to be used after this call?